beacon-chain/blockchain/blob_data_column_notifier.go

```diff
@@ -0,0 +1,113 @@
+	// 1. p2p layer receiving data columns from blob column subnets
+	// 2. sampling requests data_column_sidecars_by_root
+	// and updates the internal state of the notifier to reflect that the data column is available.
+	ReceiveBlobDataColumn(blockRoot [32]byte, columnIndex uint64)
+
+	// MissingDataColumns returns the list of missing data columns for the given blockRoot.
+	MissingDataColumns(blockRoot [32]byte) []uint64
+}
+
+// NewBlobDataColumnNotifier creates a new blobDataColumnNotifier.
+func NewBlobDataColumnNotifier() blobAvailabilityNotifierMap {
+	return &blobDataColumnAvailabilityNotifierMap{
+		notifiers: make(map[[32]byte]chan struct{}),
+		seenIndex: make(map[[32]byte]map[uint64]bool),
+	}
+}
+
+var _ blobAvailabilityNotifierMap = &blobDataColumnAvailabilityNotifierMap{}
+
+type blobDataColumnAvailabilityNotifierMap struct {
+	sync.RWMutex
+
+	// notifiers is a map of blockRoot to a channel that will be closed when a new blob is available for that blockRoot.
+	notifiers map[[32]byte]chan struct{}
+
+	// seenIndex is a map of blockRoot to a set of column indexes that have been seen for that blockRoot.
+	seenIndex map[[32]byte]map[uint64]bool
+}
+
+// MissingDataColumns implements blobAvailabilityNotifierMap.
+func (bn *blobDataColumnAvailabilityNotifierMap) MissingDataColumns(blockRoot [32]byte) []uint64 {
+	bn.RLock()
+	defer bn.RUnlock()
+
+	missing := make([]uint64, fieldparams.NumberOfColumns)
+	seen := bn.seenIndex[blockRoot]
+	for i := uint64(0); i < fieldparams.NumberOfColumns; i++ {
+		if seen[i] {
+			continue
+		}
+		missing = append(missing, i)
+	}
+
+	return missing
+}
+
+// ReceiveBlobDataColumn implements blobDataColumnNotifier. It notifies the notifier that a new blob data column is available for the given blockRoot.
+func (bn *blobDataColumnAvailabilityNotifierMap) ReceiveBlobDataColumn(blockRoot [32]byte, columnIndex uint64) {
+	if columnIndex >= fieldparams.NumberOfColumns {
+		return
+	}
+
+	bn.Lock()
+	defer bn.Unlock()
+	seen := bn.seenIndex[blockRoot]
+	if seen == nil {
+		seen = make(map[uint64]bool)
+		bn.seenIndex[blockRoot] = seen
+	}
+	if seen[columnIndex] {
+		return
+	}
+	seen[columnIndex] = true
+
+	if !bn.isDataAvailable(seen) {
+		return
+	}
+
+	notifier, ok := bn.notifiers[blockRoot]
+	if !ok {
+		notifier = make(chan struct{})
+		bn.notifiers[blockRoot] = notifier
+	}
+
+	notifier <- struct{}{}
+	delete(bn.notifiers, blockRoot)
+	delete(bn.seenIndex, blockRoot)
+}
+
+// DataAvailable implements blobDataColumnNotifier. It returns a channel that will be closed when a new blob is available for the given blockRoot.
+func (bn *blobDataColumnAvailabilityNotifierMap) DataAvailable(blockRoot [32]byte) <-chan struct{} {
+	bn.Lock()
+	defer bn.Unlock()
+
+	notifier, ok := bn.notifiers[blockRoot]
+	if !ok {
+		notifier = make(chan struct{})
+		bn.notifiers[blockRoot] = notifier
+	}
+
+	return notifier
+}
+
+func (bn *blobDataColumnAvailabilityNotifierMap) isDataAvailable(seenColumns map[uint64]bool) bool {
```

Comment:
@nalepae before I spend more time on this, would love to get your opinion regarding sampling timing.

Currently the code 
1. `is_data_available` returns `true` when all the columns it's supposed to custody are stored, it does not taken sampling into account, meaning if sampling fails, the block is actually not available.
2. does sampling asynchronously with `go s.dataColumnSampling()` when `statefeed.BlockProcessed` event is received.

I'm not super familiar with the codebase, but my thought is that this is a timing game, depending on how fork choice rule is designed (currently not clear) in the spec
1. we'll need to either make data sampling synchronous in the block transition process so that we could have good enough confidence on whether the data is available or not, in this case, we could potentially kick off separate goroutine for sampling as soon as we received the first data column for a new block root.
2. Or we could keep current structure, make sure we've downloaded all the data we're supposed to custody first, then do the sampling, but we cannot do sampling asynchronously like current implementation, there needs to be a `wait` method in the `ReceiveBlock` function that waits for sampling success before applying state transitions, etc

Curious your thoughts on this?

Reply:
You're right. Currently:

- `isDataAvailable` does not take any sampling results into account, and
- We do run sampling, but the results are only displayed in some logs. How peer sampling will influence the fork choice rule is still under discussion, and the corresponding specification has not yet been merged.

You can look at this [peerDAS design progress](https://notes.ethereum.org/sFGCCOhYTjGbVH_lzItJnA?view) for more information.